software/bootcamp-code/main.py

```diff
@@ -33,61 +108,257 @@ def main_loop():
+            elif state==State.TEST_THROWER:
+                omni_motion.move(0, 0, 0, testing_thrower_speed)
+                print("testing thrower")
+                processedData=processor.process_frame(aligned_depth=False)
+                basket_depth = get_depth(processedData.depth_frame, 0, basket_to_throw.x)
+                print(basket_depth)
+                #print(processor.process_frame(aligned_depth=False).basket_b.distance)
+
+            elif state==State.FIND_FURTHEST_BASKET:
+                # rotate and find both the baskets, when found, drive to the furthest
+                if processedData.basket_b.exists:
+                    blue_basket_depth = get_depth(processedData.depth_frame, (processedData.basket_b.y+100), processedData.basket_b.x)
+                    if blue_basket_depth>basket_blue_d: basket_blue_d=blue_basket_depth
+                elif processedData.basket_m.exists:
+                    magn_basket_depth = get_depth(processedData.depth_frame, (processedData.basket_m.y+100), processedData.basket_m.x)
+                    if magn_basket_depth>basket_magn_d: basket_magn_d=magn_basket_depth
+
+                # compare and see which is furthest
+                if basket_blue_d>=basket_magn_d: furthest_basket=BasketColor.BLUE
+                elif basket_magn_d>basket_blue_d: furthest_basket=BasketColor.MAGENTA
                 
-                # The destination coordinates are the difference between the ball location and desired location.
-                dest_x = ball_desired_x - processedData.balls[0].x
-                dest_y = ball_desired_y - processedData.balls[0].y
-                # Normalize destination in the [-1;1] range and and then multiply it with motor max speed.
-                # (then the driving is proportional)
-                # Speed range for motors is 48 - 2047, we use 100 for max motor speed at the moment.
-                x_speed = (dest_x/cam.rgb_width) * 100
-                y_speed = (dest_y/cam.rgb_height) * 100
-                #rotation = 0
-                # Drive towards a ball:
-                    # side speed is x_speed
-                    # forward speed is y_speed
-                    # rotation if you want to turn
-                #print("X: ", x_speed, "Y: ", y_speed)
-                omni_motion.move(x_speed, y_speed, 0)
+                print(furthest_basket, basket_blue_d, basket_magn_d)
+                if basket_blue_d>0 and basket_magn_d>0 and basket_to_drive.exists:
+                    state=State.DRIVE_TO_OP_BASKET
+                    continue
+                omni_motion.move(0, 0, find_rotation_speed/2)
+
+            # State to drive to the furthest basket when in the find_ball state the ball has not been found for some time.
+            elif state==State.DRIVE_TO_OP_BASKET:
+                basket_depth = get_depth(processedData.depth_frame, 0, basket_to_drive.x)
+                if basket_depth<250: # if basket is closer than 20cm
```

Comment:
I suggest more state exiting strategies, for example when it seems a ball for a few frames.